lib/posix-process/signal/signal.c

```diff
@@ -175,6 +185,126 @@ void pprocess_signal_sigaction_clear(struct uk_sigaction *ks)
 	uk_sigemptyset(&ks->ks_mask);
 }
 
+int pprocess_signal_send(struct posix_process *proc, int signum,
+			 siginfo_t *siginfo)
+{
+	struct uk_signal *sig;
+	int rc;
+
+	sig = uk_signal_alloc(proc->_a);
+	if (siginfo)
+		set_siginfo_sigqueue(signum, &sig->siginfo, siginfo);
+	else
+		set_siginfo_kill(signum, &sig->siginfo);
+
+	rc = pprocess_signal_queue_p(sig, proc);
+	if (unlikely(rc)) {
+		uk_signal_free(proc->_a, sig);
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+int pprocess_signal_process_do(pid_t pid, int signum, siginfo_t *siginfo)
+{
+	struct posix_process *pproc = NULL;
+	struct posix_thread *thread, *threadn;
+	int rc;
+
+	if (unlikely(signum != 0 && !IS_VALID(signum)))
+		return -EINVAL;
+
+	if (unlikely(signum == SIGSTOP || signum == SIGCONT)) {
+		uk_pr_warn("Process stop / resume not supported. Ignoring\n");
+		return 0;
+	}
+
+	/* pid == 0 -> Send the signal to every process in the process group
+	 *             of the calling process.
+	 * pid < -1 -> Send the signal to every process in the process group
+	 *             with ID equal to -pid.
+	 */
+	if (!pid || pid < -1) {
+		uk_pr_warn("pgroups not supported, delivering to every process\n");
+		pid = -1;
+	}
+
+	/* pid == -1 -> Send the signal to every process for which the
+	 *              calling process has permissions to signal.
+	 */
+	if (pid == -1) {
+		uk_process_foreach(pproc) {
+			if (!signal_check_perm(pproc))
+				continue;
+
+			if (IS_IGNORED(pproc, signum))
+				continue;
+
+			rc = pprocess_signal_send(pproc, signum, siginfo);
+			if (unlikely(rc))
+				return rc;
+		}
+		return 0;
+	}
+
+	/* Default: Send to process with pid */
+	pproc = pid2pprocess(pid);
+	if (unlikely(!pproc))
+		return -ESRCH;
+
+	if (unlikely(signal_check_perm(pproc) == false))
+		return -EPERM;
+
+	/* If signum == 0 do only pid exist & permissions check */
+	if (!signum)
+		return 0;
+
+	/* If this signal is currently ignored, don't even try */
+	if (IS_IGNORED(pproc, signum))
+		return 0;
+
+	rc = pprocess_signal_send(pproc, signum, siginfo);
+	if (unlikely(rc))
+		return rc;
+
+	/* Wake up any threads that may be paused */
+	uk_process_foreach_pthread(pproc, thread, threadn) {
+		if (thread->state == POSIX_THREAD_BLOCKED_SIGNAL)
+			uk_semaphore_up(&thread->signal->deliver_semaphore);
+	}
+
+	return 0;
+}
+
+int pprocess_signal_thread_do(int tid, int signum, siginfo_t *siginfo)
+{
+	struct posix_thread *pthread;
+	struct uk_signal *sig;
+	int rc;
+
+	pthread = tid2pthread(tid);
+	if (unlikely(!pthread))
+		return -EINVAL;
+
+	sig = uk_signal_alloc(pthread->_a);
+	if (siginfo)
+		set_siginfo_sigqueue(signum, &sig->siginfo, siginfo);
+	else
+		set_siginfo_kill(signum, &sig->siginfo);
+
+	rc = pprocess_signal_queue_t(sig, pthread);
+	if (unlikely(rc)) {
+		uk_signal_free(pthread->_a, sig);
+		return -EAGAIN;
```

Comment:
Should we maybe propagate `rc` instead? `pprocess_signal_queue_t` can return `EAGAIN` on its own.

Reply:
Yes, most probably this is a remaining from a previous iteration.